[DAG] Use copysign in frem power-2 fold.

davemgreen opened the pull request:
As a small addition to #91148, this uses copysign to produce the correct sign for zero when converting frem to div/trunc/mul when we do not know that the input is positive (and we care about sign bits). The copysign lets us get the sign of zero correct.

In testing, the only case this produced different results than fmod was:
frem -inf, 4.0 -> nan vs -nan

jayfoad commented:
> when we do not know that the input is positive (or we care about sign bits)

"**and** we care about sign bits", surely?

davemgreen committed:
[DAG] Use copysign in frem power-2 fold.

As a small addition to #91148, this uses copysign to produce the correct sign
for zero when converting frem to div/trunc/mul when we do not know that the
input is positive (and we care about sign bits). The copysign lets us get the
sign of zero correct.

In testing, the only case this produced different results thant fmod was:
frem -inf, 4.0 -> nan vs -nan

davemgreen commented:
> > when we do not know that the input is positive (or we care about sign bits)
> 
> "**and** we care about sign bits", surely?

Sure, sounds good. The original was even more wrong, and I messed up the reference commit too.

davemgreen commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -17383,15 +17383,22 @@ SDValue DAGCombiner::visitFREM(SDNode *N) {
+      DAG.isKnownToBeAPowerOfTwoFP(N1)) {
+    bool NeedsCopySign =
+        !Flags.hasNoSignedZeros() && !DAG.cannotBeOrderedNegativeFP(N0);
+    if (!NeedsCopySign || TLI.isOperationLegalOrCustom(ISD::FCOPYSIGN, VT)) {
```
I was aiming for either knowing that the COPYSIGN isn't necessary, or checking that the lowering was going to be at least somewhat sensible. I think you might be right though, if we've already checked legality the conditions above we can assume the copysign is fine too.

davemgreen committed:
Remove COPYSIGN legality check
